Upgrade-to-vite

mayank1513 opened the pull request:
- Replace react-scripts with vite for better build and live reload speed
  - better dx
  - modern tooling
- Set up vitest for unit testing
- Updated dependencies to latest
- Setup ESLint as per vite
- Set up prettier for consistent code formatting
- Refactor for better readability and maintainability 
- Fix lint issues

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/eclipsesource/jsonforms-react-seed?pullRequest=119) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/eclipsesource/jsonforms-react-seed?pullRequest=119) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/eclipsesource/jsonforms-react-seed?pullRequest=119) it.</sub>

lucas-koehler commented on package.json:
```diff
@@ -1,38 +1,41 @@
+    "@mui/lab": "5.0.0-alpha.170",
+    "@mui/material": "^5.15.19",
+    "@mui/x-date-pickers": "^6.x",
+    "@types/node": "^20.14.2",
+    "@types/react": "^18.3.3",
+    "@types/react-dom": "^18.3.0",
```
Can we move these to `devDependencies` please?

mayank1513 replied:
Sure

lucas-koehler commented on .gitignore:
```diff
@@ -31,6 +31,7 @@ bower_components
 
 # Compiled binary addons (http://nodejs.org/api/addons.html)
 build/Release
+dist
```
As the build is placed in the `build` directory, I believe we don't need this. Or am I missing something?

lucas-koehler reviewed (changes requested):
Hi @mayank1513 ,

thanks for this contribution :heart: . Upgrading the seed app from CRA to ViteJS is certainly a good idea :+1: 

I have two remarks inline and noticed three further things:

- Linting no longer works. The whole `"eslintConfig"` property in the package.json is invalid as the referenced configs are part of CRA and, thus, no longer present. This should be removed. With this, there is no more lint config. I suppose it would make sense to use a vitejs default lint setup for typescript. Could you add this?
- For me the tests (via `npm run test`) cannot be executed because a dependency for `jsdom` is missing. We should add it as a dev dependency
- Optional: The build complains that the deprecated CJS (common js) build is used. We could adapt the package.json and tsconfig.json to use ESM.

mayank1513 commented:
> Hi @mayank1513 ,
> 
> thanks for this contribution ❤️ . Upgrading the seed app from CRA to ViteJS is certainly a good idea 👍
> 
> I have two remarks inline and noticed three further things:
> 
>     * Linting no longer works. The whole `"eslintConfig"` property in the package.json is invalid as the referenced configs are part of CRA and, thus, no longer present. This should be removed. With this, there is no more lint config. I suppose it would make sense to use a vitejs default lint setup for typescript. Could you add this?
> 
>     * For me the tests (via `npm run test`) cannot be executed because a dependency for `jsdom` is missing. We should add it as a dev dependency
> 
>     * Optional: The build complains that the deprecated CJS (common js) build is used. We could adapt the package.json and tsconfig.json to use ESM.

Thank you for your observations. Probably I did not notice these as I was using yarn and pnpm locally. I have fixed the issues you observed.

mayank1513 commented:
Initially there was issue with even launching the Cypress test. I fixed
that. However the tests were not passing in the GitHub action - it was
throwing timeout error.

Can you help debug and fix this?

On Mon, 10 Jun, 2024, 8:52 pm Lucas Koehler, ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> Hi @mayank1513 <https://github.com/mayank1513>, thanks for the updates!
>
> I only have a few minor comments inline. I also noticed that npm run
> cypress:ci does not seem to work and the tests fail. do the cypress tests
> work for you?
> ------------------------------
>
> On .tkb
> <https://github.com/eclipsesource/jsonforms-react-seed/pull/119#discussion_r1633404528>
> :
>
> i think this was commited by mistake?
> ------------------------------
>
> In .vscode/extensions.json
> <https://github.com/eclipsesource/jsonforms-react-seed/pull/119#discussion_r1633404967>
> :
>
> > @@ -5,7 +5,7 @@
>    // List of extensions which should be recommended for users of this workspace.
>    "recommendations": [
>      "esbenp.prettier-vscode",
> -    "ms-vscode.vscode-typescript-tslint-plugin"
> +    "mayank1513.trello-kanban-task-board"
>
> this should not be added here.
> ------------------------------
>
> In .vscode/settings.json
> <https://github.com/eclipsesource/jsonforms-react-seed/pull/119#discussion_r1633405582>
> :
>
> > \ No newline at end of file
> +  "editor.formatOnSave": true,
> +  "editor.defaultFormatter": "esbenp.prettier-vscode",
> +  "prettier.jsxSingleQuote": true,
> +  "prettier.singleQuote": true,
> +  "mayank1513.trello-kanban.Workspace.filePath": ".tkb"
>
> we should not need this. please remove :)
> ------------------------------
>
> On src/components/schema.json
> <https://github.com/eclipsesource/jsonforms-react-seed/pull/119#discussion_r1633414071>
> :
>
> The schema.json is not a component and, thus, should not be moved to the
> components directory. The old location directly in src was fine. This makes
> it also slightly easier to find for playing around
> ------------------------------
>
> On src/components/uischema.json
> <https://github.com/eclipsesource/jsonforms-react-seed/pull/119#discussion_r1633414496>
> :
>
> Same as for schema.json. This should stay in src
> ------------------------------
>
> On package.json
> <https://github.com/eclipsesource/jsonforms-react-seed/pull/119#discussion_r1633418550>
> :
>
> As the starter is now built and implemented as ESM, we should add "type":
> "module".
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipsesource/jsonforms-react-seed/pull/119#pullrequestreview-2108057255>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGCR4VN6OYH6NF24G22DQ6DZGXACBAVCNFSM6AAAAABI5SOJY6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCMBYGA2TOMRVGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mayank1513 commented on .vscode/extensions.json:
```diff
@@ -5,7 +5,7 @@
   "recommendations": [
     "esbenp.prettier-vscode",
-    "ms-vscode.vscode-typescript-tslint-plugin"
+    "mayank1513.trello-kanban-task-board"
```
done

mayank1513 commented:
Cypress is failing because this element does not exist in the DOM - `[id="#/properties/recurrence"] > div`

mayank1513 commented:
Fixed Cypress tests...

lucas-koehler reviewed (approved):
Hi @mayank1513 , thank you very much for the updates and the whole rework! Changes LGTM now :sparkles: 

I added commits for three small changes:
- Fix one lint error in the cypress config
- Execute CI in pull requests and add linting to CI
- Regenerate the package-lock.json: We always do that for external/unknown contributors as a general security precaution.
